src/Program2.cs

```diff
@@ -0,0 +1,526 @@
+
+//class Doc
+//{
+//    async IAsyncEnumerable<int> Illustration()
+//    {
+//        // call to WaitForNext will start all the code until state 1
+//        Console.Write("abc");
+//        await Slow(); // state 0, machine automatically continues
+
+//        Console.Write("def");
+//        await Slow(); // state 2, machine automatically continues
+
+//        Console.Write("ghi");
+//        yield return 42; // state 1, promise of a value is fulfilled, WaitForNext completes with true, promise is left as signal of available value
+//                         // first call to TryGetNext notices promise from running machine, clears it, returns value and true, moves to state 3
+//                         // second call to TryGetNext prints "jkl", has no promise to fulfill, returns true, moves to state 3
+
+//        Console.Write("jkl");
+//        yield return 43; // state 3
+//                         // third call to TryGetNext prints "mno", has no promise to fulfill, returns false, moves to state 4, lets machine run asynchronously
+
+//        Console.Write("mno");
+//        await Slow(); // state 4, machine automatically continues
+//                      // call to WaitForNext in parallel with code executing already
+
+//        Console.Write("pqr");
+//        await Slow(); // state 6, machine automatically continues
+
+//        Console.Write("stu");
+//        yield return 44; // state 5, promise of a value is fulfilled, WaitForNext completes with true
+//                         // fourth call to TryGetNext returns true, moves to state -2
+//                         // fifth call to TryGetNext returns false
+//                         // call to WaitForNext returns false
+//    }
+
+//    Task Slow() => throw null;
+//}
+
+class Program
+{
+    static async Task Main()
+    {
+        IAsyncEnumerable<int> test = AsyncIterator();
+        IAsyncEnumerator<int> enumerable = test.GetAsyncEnumerator();
+
+        //foreach await (var value in enumerable)
+        //{
+        //    Console.WriteLine(value)
+        //}
+        while (await enumerable.WaitForNextAsync())
+        {
+            while (true)
+            {
+                int value = enumerable.TryGetNext(out bool success);
+                if (!success) goto outer_loop_continue;
+
+                Console.WriteLine(value);
+            }
+            outer_loop_continue:;
+        }
+    }
+
+    static IAsyncEnumerable<int> AsyncIterator()
+    {
+        var stateMachine = new Unprounouncable();
+        stateMachine.State = StateMachineStates.NotStartedStateMachine;
+        stateMachine.Builder = AsyncTaskMethodBuilder.Create();
+        stateMachine._valueOrEndPromise = new ManualResetValueTaskSourceLogic<bool>(stateMachine);
+        stateMachine._promiseIsActive = true;
+
+        // note: we don't start the machine.
+        return stateMachine;
+    }
+
+    // TODO what happens to exceptions? (should they bubble into value promise, or anywhere else?)
```

Comment:
I think they should be stored in the ValueTask returned from WaitForNextAsync.  If an exception has been captured, TryGetNext would return a false Boolean value, and the next call to WaitForNextAsync would return the faulted ValueTask.